src/Dotnet.Script.DependencyModel.Nuget/NuGetSourceReferenceResolver.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis;
+
+namespace Dotnet.Script.DependencyModel.NuGet
+{
+    /// <summary>
+    /// A <see cref="SourceReferenceResolver"/> decorator that handles
+    /// references to NuGet packages in scripts.  
+    /// </summary>
+    public class NuGetSourceReferenceResolver : SourceReferenceResolver
+    {
+        private readonly SourceReferenceResolver _sourceReferenceResolver;
+        private readonly IDictionary<string, IReadOnlyList<string>> _scriptMap;
+        private static readonly Regex PackageNameMatcher = new Regex(@"\s*nuget\s*:\s*(.*)\s*,", RegexOptions.Compiled | RegexOptions.IgnoreCase); 
+
+        public NuGetSourceReferenceResolver(SourceReferenceResolver sourceReferenceResolver, IDictionary<string, IReadOnlyList<string>> scriptMap)
+        {
+            _sourceReferenceResolver = sourceReferenceResolver;
+            _scriptMap = scriptMap;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj.GetType() != GetType()) return false;
+            return _sourceReferenceResolver.Equals(obj);
+        }
+
+        public override int GetHashCode()
+        {
+            return _sourceReferenceResolver.GetHashCode();
+        }
+
+        public override string NormalizePath(string path, string baseFilePath)
+        {
+            if (path.StartsWith("nuget", StringComparison.OrdinalIgnoreCase))
+            {
+                return path;
+            }
+
+            return _sourceReferenceResolver.NormalizePath(path, baseFilePath);
+        }
+
+        public override string ResolveReference(string path, string baseFilePath)
+        {
+            if (path.StartsWith("nuget", StringComparison.OrdinalIgnoreCase))
+            {
+                var packageName = PackageNameMatcher.Match(path).Groups[1].Value;
+                var scripts = _scriptMap[packageName];
```

Comment:
this will unfortunately have the same issue as the nuget resolver for `#r`, in that it will not work with OmniSharp in real time (as you type) but only on initial file opening.

We will need to look at injecting sort of a state machine into these resolver and if resolve the references from the state machine; if not found there, we run `dotnet restore` again. Not very elegant but I don't think there is a better way to do it with the current Roslyn extensibility points